core/src/components/datetime/utils/manipulation.ts

```diff
@@ -472,25 +472,64 @@ export const getClosestValidDate = ({
   const copyParts = clampDate({ ...refParts, dayOfWeek: undefined }, minParts, maxParts);
 
   if (monthValues !== undefined) {
-    copyParts.month = findClosestValue(month, monthValues);
+    // Filters out months that are out of the min/max bounds
+    const filteredMonths = monthValues.filter(
+      (month) =>
+        !(minParts !== undefined && month < minParts.month!) && !(maxParts !== undefined && month > maxParts.month!)
+    );
+    copyParts.month = findClosestValue(month, filteredMonths);
   }
 
   // Day is nullable but cannot be undefined
   if (day !== null && dayValues !== undefined) {
-    copyParts.day = findClosestValue(day, dayValues);
+    // Filters out days that are out of the min/max bounds
+    const filteredDays = dayValues.filter((day) => {
+      if (minParts && isBefore({ ...copyParts, day }, minParts)) {
+        return false;
+      }
+      if (maxParts && isAfter({ ...copyParts, day }, maxParts)) {
+        return false;
+      }
+      return true;
+    });
+    copyParts.day = findClosestValue(day, filteredDays);
   }
 
   if (yearValues !== undefined) {
-    copyParts.year = findClosestValue(year, yearValues);
+    // Filters out years that are out of the min/max bounds
+    const filteredYears = yearValues.filter(
+      (year) => !(minParts !== undefined && year < minParts.year!) && !(maxParts !== undefined && year > maxParts.year!)
+    );
+    copyParts.year = findClosestValue(year, filteredYears);
   }
 
   if (hour !== undefined && hourValues !== undefined) {
-    copyParts.hour = findClosestValue(hour, hourValues);
+    // Filters out hours that are out of the min/max bounds
+    const filteredHours = hourValues.filter((hour) => {
+      if (minParts?.hour !== undefined && isSameDay(copyParts, minParts) && hour < minParts.hour) {
+        return false;
+      }
+      if (maxParts?.hour !== undefined && isSameDay(copyParts, maxParts) && hour > maxParts.hour) {
+        return false;
+      }
+      return true;
+    });
+    copyParts.hour = findClosestValue(hour, filteredHours);
     copyParts.ampm = parseAmPm(copyParts.hour);
   }
 
   if (minute !== undefined && minuteValues !== undefined) {
-    copyParts.minute = findClosestValue(minute, minuteValues);
+    // Filters out minutes that are out of the min/max bounds
+    const filteredMinutes = minuteValues.filter((minute) => {
+      if (minParts?.minute !== undefined && isSameDay(copyParts, minParts) && minute < minParts.minute) {
+        return false;
+      }
+      if (maxParts?.minute !== undefined && isSameDay(copyParts, maxParts) && minute > maxParts.minute) {
```

Comment:
This should only apply if the copyParts.hour matches the min/max hour like in my original example. `isSameDay` does not automatically account for `hour`. Otherwise, the minute value will be adjusted when it does not need to be.

Test case:

```js
// April 2, 2020 9:16 AM
const refParts = { month: 4, day: 2, year: 2020, hour: 9, minute: 16 };
// September 10, 2021 10:10 AM
const minParts = { month: 9, day: 10, year: 2021, hour: 10, minute: 10 };
// September 10, 2021 11:15 AM
const maxParts = { month: 9, day: 10, year: 2021, hour: 11, minute: 15 };

// September 10, 2021 10:15 AM
const expected = {
  month: 9,
  day: 10,
  year: 2021,
  hour: 10,
  minute: 16,
  ampm: 'am',
  dayOfWeek: undefined,
};

expect(
  getClosestValidDate({
    refParts,
    monthValues: [4, 9, 11],
    dayValues: [10, 12, 13, 14],
    yearValues: [2020, 2021, 2023],
    hourValues: [9, 10, 11],
    minuteValues: [10, 15, 16],
    minParts,
    maxParts,
  })
).toEqual(expected);
```